src/ws/message.rs

```diff
@@ -0,0 +1,197 @@
+use bytes::{Buf, Bytes, TryGetError};
+use bytestring::ByteString;
+use serde::Serialize;
+use serde::de::DeserializeOwned;
+use std::ops::ControlFlow;
+use tokio::sync::mpsc;
+use tokio::task::coop;
+use tokio_websockets::proto::ProtocolError;
+
+use super::error::ErrorKind;
+use crate::error::Error;
+use crate::payload::{Payload, deserialize_json};
+
+pub use tokio_websockets::CloseCode;
+
+pub(super) type Sender = mpsc::Sender<Message>;
+pub(super) type Receiver = mpsc::Receiver<Message>;
+
+pub struct Channel(Sender, Receiver);
+
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum Message {
+    Binary(Bytes),
+    Close(Option<(CloseCode, Option<ByteString>)>),
+    Text(ByteString),
+}
+
+impl Channel {
+    pub(super) fn new(tx: Sender, rx: Receiver) -> Self {
+        Self(tx, rx)
+    }
+
+    pub async fn send(&mut self, message: impl Into<Message>) -> super::Result<()> {
+        if self.0.send(message.into()).await.is_err() {
+            Err(ControlFlow::Break(ErrorKind::CLOSED.into()))
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn recv(&mut self) -> impl Future<Output = Option<Message>> {
+        coop::unconstrained(self.1.recv())
```

Comment:
Both of the mpsc receivers are unconstrained. This keeps the channel latency near-zero.

The inner loop of the start fn in the ws::upgrade module keeps the websocket handler cooperative while allowing a buffer size of 1. The channel is more like a callback than a form of back pressure.

This should scale nicely. It keeps the user code removed from the actual socket without the latency and queuing associated with a channel.